NitroxServer/GameLogic/Bases/BaseManager.cs

```diff
@@ -33,13 +35,44 @@ public List<BasePiece> GetPartiallyConstructedPieces()
                 return new List<BasePiece>(partiallyConstructedPiecesById.Values);
             }
         }
-
-        public void AddBasePiece(BasePiece basePiece)
+        public void AddBasePiece(BasePiece basePiece) 
         {
-            lock (partiallyConstructedPiecesById)
+
+            if (basePiece.TechType.Name != "BaseReinforcement" | basePiece.TechType.Name != "BaseHatch" | basePiece.TechType.Name != "BaseWindow")
             {
+                lock (partiallyConstructedPiecesById)
+                {
+                    partiallyConstructedPiecesById.Add(basePiece.Id, basePiece);
+                }
+            }
+
+            if (basePiece.TechType.Name == "BaseWindow")
+            {
+                basePiece.ConstructionCompleted = false;
+                partiallyConstructedPiecesById.Add(basePiece.Id, basePiece);
+                basePiece.ConstructionAmount = 1.0f;
+                basePiece.ConstructionCompleted = true;
+            }
+            if (basePiece.TechType.Name == "BaseHatch")
+            {
+                basePiece.ConstructionCompleted = false;
                 partiallyConstructedPiecesById.Add(basePiece.Id, basePiece);
+                basePiece.ConstructionAmount = 1.0f;
+                basePiece.ConstructionCompleted = true;
             }
+            if (basePiece.TechType.Name == "BaseReinforcement")
+            {
+                basePiece.ConstructionCompleted = false;
+                partiallyConstructedPiecesById.Add(basePiece.Id, basePiece);
+                basePiece.ConstructionAmount = 1.0f;
+                basePiece.ConstructionCompleted = true;
+            }
+            if (basePiece.ConstructionAmount > 0.4f)
+            {
+                basePiece.ConstructionAmount = 1.0f;
+                basePiece.ConstructionCompleted = true;
+            }
+
```

Comment:
Maybe add some comments to the code in order to allows us to understand what the idea you had behind that code